addon/-private/system/many-array.js

```diff
@@ -87,15 +87,39 @@ export default Ember.Object.extend(Ember.MutableArray, Ember.Evented, {
     );
     toSet = toSet.concat(newRecords);
     var oldLength = this.length;
-    this.arrayContentWillChange(0, this.length, toSet.length);
-    // It’s possible the parent side of the relationship may have been unloaded by this point
-    if (_objectIsAlive(this)) {
-      this.set('length', toSet.length);
+    var newLength = toSet.length;
+
+    var shortestLength = Math.min(oldLength, newLength);
+
+    var firstChangeIndex = -1; // -1 signifies no changes
+    // find the first change
+    var currentArray = this.currentState;
+    for (var i=0; i<shortestLength; i++) {
+      // compare each item in the array
+      if (currentArray[i] !== toSet[i]) {
+        firstChangeIndex = i;
+        break;
+      }
+    }
+    if (firstChangeIndex === -1) {
+      // no change found in the matching part of the arrays
+      if (newLength !== oldLength) {
+        firstChangeIndex = shortestLength;
+      }
+    }
+    if (firstChangeIndex !== -1) {
+      // we found a change
+      var added = newLength - firstChangeIndex;
```

Comment:
what happens if we have non-contiguous changes, e.g. the collection was replaced and some members of the previous set and the current set are the same.

```js
[model#1, model#2, model#3, model#4, model#10]
```

turns into:

```js
[model#1, model#5, model#10]
```